services/QuillLMS/engines/evidence/spec/services/evidence/gen_ai/on_topic/data_fetcher_spec.rb

```diff
@@ -0,0 +1,148 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'csv'
+
+RSpec.describe Evidence::GenAI::OnTopic::DataFetcher, type: :service do
+  describe 'constants' do
+    it 'has a default file' do
+      expect(described_class::DEFAULT_FILE).to eq('full_data.csv')
+    end
+
+    it 'defines valid types' do
+      expect(described_class::TYPES).to contain_exactly('test', 'train')
+      expect(described_class::TYPE_TEST).to eq('test')
+      expect(described_class::TYPE_TRAIN).to eq('train')
+    end
+  end
+
+  describe 'initialization' do
+    context 'with valid type' do
+      it 'allows test type' do
+        expect { described_class.new(type: 'test') }.not_to raise_error
+      end
+
+      it 'allows train type' do
+        expect { described_class.new(type: 'train') }.not_to raise_error
+      end
+    end
+
+    context 'with invalid type' do
+      it 'raises an error' do
+        expect { described_class.new(type: 'invalid') }.to raise_error(StandardError)
+      end
+    end
+
+    context 'with optional parameters' do
+      let(:fetcher) { described_class.new(type: 'test', prompt_id: 123, on_topic: true, limit: 5, file: 'custom.csv') }
+
+      it 'sets all attributes correctly' do
+        expect(fetcher.type).to eq('test')
+        expect(fetcher.prompt_id).to eq(123)
+        expect(fetcher.on_topic).to be(true)
+        expect(fetcher.limit).to eq(5)
+        expect(fetcher.file).to eq('custom.csv')
+      end
```

Comment:
This is a take-or-leave item since I think my approach here is a lot more verbose, and may not be worth it, but I'm always nervous when there's a hard-coded value in setup like a `let` that's then hard-coded again in an `expect`.  I think it's more clear that the input and output are connected when it's explicit:
```ruby
let(:type) { 'test' }
...
let(:fetcher) { described_class.new(type:, ...) }

it ...
  expect(fetcher.type).to eq(type)
  ...
end
```

Reply:
Fair, I'll clean these up a bit.